backend/app/db/dataset/version.py

```diff
@@ -165,3 +173,298 @@ async def _freeze_dataset_visualization(
+                    )
+                    is None
+                ):
+                    fs.remove_object(settings.MINIO_BUCKET_NAME, str(vis_data.id))
+
+        return vis_config.dict()
+
+
+async def _delete_frozen_thumbnail(resource, fs):
+    async for frozen_thumbnail in ThumbnailFreezeDB.find(
+        ThumbnailFreezeDB.id == resource.thumbnail_id,
+    ):
+        # delete mongo document
+        await frozen_thumbnail.delete()
+
+        if fs is not None:
+            # if thumbnail is not used by other frozen dataset, delete the raw bytes
+            if (
+                await ThumbnailFreezeDB.find_one(
+                    ThumbnailFreezeDB.origin_id
+                    == PydanticObjectId(frozen_thumbnail.origin_id)
+                )
+                is None
+            ) and (
+                await ThumbnailDB.find_one(ThumbnailDB.id != frozen_thumbnail.origin_id)
+                is None
+            ):
+                fs.remove_object(
+                    settings.MINIO_BUCKET_NAME, str(frozen_thumbnail.origin_id)
+                )
+
+        return frozen_thumbnail.dict()
+
+
+async def _delete_thumbnail(resource, fs):
+    async for thumbnail in ThumbnailDB.find(
+        ThumbnailDB.id == resource.thumbnail_id,
+    ):
+        # delete mongo document
+        await thumbnail.delete()
+
+        if fs is not None:
+            # if thumbnail is not used by other frozen dataset, delete the raw bytes
+            if (
+                await ThumbnailFreezeDB.find_one(
+                    ThumbnailFreezeDB.origin_id == PydanticObjectId(thumbnail.id)
+                )
+                is None
+            ):
+                fs.remove_object(settings.MINIO_BUCKET_NAME, str(thumbnail.id))
+
+        return thumbnail.dict()
+
+
+async def _delete_file(file, fs):
+    # delete mongo
+    await file.delete()
+
+    if fs is not None:
+        # delete file raw bytes if not used anywhere else
+        if (
+            await FileFreezeDB.find_one(
+                FileFreezeDB.origin_id == PydanticObjectId(file.id)
+            )
+            is None
+        ):
+            fs.remove_object(settings.MINIO_BUCKET_NAME, str(file.id))
+
+    return file.dict()
+
+
+async def _delete_frozen_file(frozen_file, fs):
+    # delete mongo
+    await frozen_file.delete()
+
+    if fs is not None:
+        # delete file raw bytes if not used anywhere else, check both frozen and well as current
+        if (
+            await FileFreezeDB.find_one(
+                FileFreezeDB.origin_id == PydanticObjectId(frozen_file.origin_id)
+            )
+            is None
+        ) and (
+            await FileDB.find_one(FileDB.id == PydanticObjectId(frozen_file.origin_id))
+        ) is None:
+            fs.remove_object(settings.MINIO_BUCKET_NAME, str(frozen_file.origin_id))
+
+    return frozen_file.dict()
+
+
+async def _delete_frozen_dataset(frozen_dataset, fs, hard_delete=False):
```

Comment:
when is hard_delete set to True?

Reply:
When deleting the whole dataset (Latest), it will be set to true to ensure all the bytes in minio are gone as well. 
In the code, the place it is set to true is on the Delete endpoint:

```
@router.delete("/{dataset_id}")
async def delete_dataset(
...
 # find associate frozen datasets and delete them iteratively
        async for frozen_dataset in DatasetFreezeDB.find(
            DatasetFreezeDB.origin_id == PydanticObjectId(dataset_id)
        ):
            await _delete_frozen_dataset(frozen_dataset, fs, hard_delete=True)
```